tools/harness-automation/autothreadharness/harness_case.py

```diff
@@ -554,28 +603,96 @@ def _test_bed(self):
             f_topo.close()
             golden_device_candidates = []
             missing_golden_devices = topo_mixed_devices[:]
-            # mapping topology config devices with devices in settings
-            for mixed_device_item in topo_mixed_devices:
-                for device_item in devices:
-                    if mixed_device_item[1] == device_item[1]:
-                        golden_device_candidates.append(device_item)
-                        devices.remove(device_item)
-                        missing_golden_devices.remove(mixed_device_item)
-                        break
-            logger.info('Golden devices in topology config file mapped in settings : %s', golden_device_candidates)
-            if len(topo_mixed_devices) != len(golden_device_candidates):
-                device_dict = dict()
-                for missing_device in missing_golden_devices:
-                    if missing_device[1] in device_dict:
-                        device_dict[missing_device[1]] += 1
-                    else:
-                        device_dict[missing_device[1]] = 1
-                logger.info('Missing Devices: %s', device_dict)
-                raise GoldenDeviceNotEnoughError()
+
+            # mapping topology config devices with golden devices and shield order in settings
+            if need_shield and shield_order:
+                for shield_item in shield_order:
+                    matched = False
+                    for mixed_device_item in topo_mixed_devices:
+                        # device need to be shielded
+                        if shield_item[1]:
+                            if 'DUT' in shield_item[0]:
+                                golden_device_candidates.append(settings.DUT2_DEVICE)
+                                dut_device = settings.DUT2_DEVICE
+                                matched = True
+                                break
+                            for device_item in shield_devices:
+                                if shield_item[0] == mixed_device_item[0] and mixed_device_item[1] == device_item[1]:
+                                    golden_device_candidates.append(device_item)
+                                    shield_devices.remove(device_item)
+                                    matched = True
+                                    break
+                        else:
```

Comment:
nit: add proper comment for multi-level nesting for better readability?

Reply:
sure